zebra/kernel_netlink.c

```diff
@@ -1080,6 +1132,288 @@ int netlink_request(struct nlsock *nl, void *req)
+		 * We received a message with the sequence number that isn't
+		 * associated with any dplane context object.
+		 */
+		if (item == NULL) {
+			zlog_debug(
+				"%s: skipping unassociated response, seq number %d NS %u",
+				__func__, h->nlmsg_seq, bth->zns->ns_id);
+			from_item = nl_batch_list_first(&(bth->items));
+			continue;
+		}
+
+		if (h->nlmsg_type == NLMSG_ERROR) {
+			int err = netlink_parse_error(nl, h, bth->zns, 0);
+
+			if (err == -1)
+				item->failure = true;
+
+			zlog_debug("%s: netlink error message seq=%d ",
+				   __func__, h->nlmsg_seq);
+			continue;
+		}
+
+		/*
+		 * If we get here then we did not receive neither the ack nor
+		 * the error and instead received some other message in an
+		 * unexpected way.
+		 */
+		zlog_debug("%s: ignoring message type 0x%04x(%s) NS %u",
+			   __func__, h->nlmsg_type,
+			   nl_msg_type_to_str(h->nlmsg_type), bth->zns->ns_id);
+	}
+
+	return 0;
+}
+
+static void nl_batch_reset(struct nl_batch *bth)
+{
+	bth->buf = nl_batch_tx_buf;
+	bth->bufsiz = sizeof(nl_batch_tx_buf);
+	bth->limit = NL_BATCH_SEND_THRESHOLD;
+
+	bth->buf_head = bth->buf;
+	bth->curlen = 0;
+	bth->msgcnt = 0;
+	bth->zns = NULL;
+
+	nl_batch_list_init(&(bth->items));
+}
+
+static void nl_batch_send(struct nl_batch *bth)
+{
+	struct nl_msg_batch_item *item;
+	bool err = false;
+
+	if (bth->curlen == 0 || bth->zns == NULL)
+		return;
+
+	if (IS_ZEBRA_DEBUG_KERNEL)
+		zlog_debug("%s: %s, batch size=%zu, msg cnt=%zu", __func__,
+			   bth->zns->nls.name, bth->curlen, bth->msgcnt);
+
+	if (netlink_send_msg(&(bth->zns->nls), bth->buf, bth->curlen) == -1)
+		err = true;
+
+	if (!err) {
+		if (nl_batch_read_resp(bth) == -1)
+			err = true;
+	}
+
+	frr_each_safe(nl_batch_list, &(bth->items), item) {
+		enum zebra_dplane_result res = ZEBRA_DPLANE_REQUEST_SUCCESS;
+
+		/*
+		 * If either sending or receiving a message batch has ended with
+		 * the error, mark all dplane requests as failed.
+		 */
+		if (item->failure || err)
+			res = ZEBRA_DPLANE_REQUEST_FAILURE;
+
+		if (!item->ignore_res)
+			dplane_ctx_set_status(item->ctx, res);
+
+		nl_batch_list_del(&(bth->items), item);
+		XFREE(MTYPE_NL_BATCH_ITEM, item);
+	}
+
+	nl_batch_reset(bth);
+}
+
+static void nl_batch_add_item(struct nl_batch *bth, int seq,
+			      struct zebra_dplane_ctx *ctx, bool ignore_res)
+{
+	struct nl_msg_batch_item *item =
+		XCALLOC(MTYPE_NL_BATCH_ITEM, sizeof(*item));
+
+	item->seq = seq;
+	item->ctx = ctx;
+	item->ignore_res = ignore_res;
+	item->failure = false;
+
+	nl_batch_list_add_tail(&(bth->items), item);
+}
+
+int netlink_batch_add_msg(struct nl_batch *bth, struct zebra_dplane_ctx *ctx,
+			  ssize_t (*msg_encoder)(struct zebra_dplane_ctx *,
+						 void *, size_t),
+			  bool extra_msg)
+{
+	int seq;
+	ssize_t size;
+	struct nlmsghdr *msgh;
+
+	if (bth->zns != NULL
+	    && bth->zns->ns_id != dplane_ctx_get_ns(ctx)->ns_id)
+		nl_batch_send(bth);
+
+	size = (*msg_encoder)(ctx, bth->buf_head, bth->bufsiz - bth->curlen);
+	if (size < 0)
```

Comment:
I added enum as you proposed, so it should be more self-descriptive now. 
`encoder` returns zero for a special kind of failure - when the message wouldn't fit in the buffer. We need to distinguish this error from others, because we can recover from this error -  simply encode this message again to the empty buffer.